Refactor: Simplify overriding wheel attributes, make wheel build deterministic

scele opened the pull request:
This is a pretty big refactoring. I have tried to separate the changes to logical commits for ease of review, but they do not necessary work correctly in isolation.

---

    Refactor generated requirements.bzl layout and make overriding attributes easier
    
    Previously, we would generate a bunch of macros and lookup tables to the
    checked-in requirements.bzl that is generated by piptool.py during
    @pip_deps//:update.  To clean things up (and to make it easier to make
    changes to these macros), generate a simple "wheels" dict instead, which
    contains all nexessary information about the resolved wheels.  Move the
    macros to requirements.bzl.tpl instead, which is generated at runtime
    and not checked into the user's repository.
    
    Remove dirty package targets, since we don't need them anymore.  Rename
    "extract_wheels" to "extract_wheel" and only accept one --whl parameter
    in "piptool resolve", since the only use case for extracting multiple
    wheels at once was the dirty wheel system.
    
    Previously, we had to expose a new pip_import attribute for every
    per-wheel customization that we whish to do: additional_runtime_deps,
    additional_buildtime_env, patch_runtime, ...  This is a bit tedious to
    maintain.  Instead, add a new "requirements_overrides" attribute, which
    accepts a dictionary in the same format as the checked-in "wheels"
    dictionary, and is merged together with that at analysis time.  This
    allows users to override any attribute of any wheel, and as we add new
    attributes to "download_or_build_wheel" and "extract_wheel" rules, they
    are automatically overridable by the new mechanism.
    
    Remove unnecessary templating from update.sh, and just generate the
    update command directly from the repository rule.

---

    Support extras and "target" argument in pip_version_proxy's requirement macro
    
    Generate new select() targets for extras, and also allow user to pass additional
    targets by adding "additional_targets" in the "requirements_overrides" dict.
    This should remove the need for "requirement_repo" macro, since the pip_version_proxy
    would now correctly proxy through things like:
    
       requirement("tensorflow", target = "cpp")
    
    Since these new targets that we use in select() are not necessary py_libraries,
    use native.alias instead of py_library targets.
    
    "binary" attribute is still left unsupported by the version proxy.

---

    Use standard patching attributes
    
    Rename "patch_runtime" to "patches", and add other standard patching
    attributes.

---

    Move build_wheel logic from skylark into piptool
    
    We will need to trigger the "build single wheel in isolation" logic
    also from "piptool resolve" when generating stable sha256s later.

---

    Make wheel build deterministic
    
    Use a fixed --build dir when invoking "pip wheel", to make debug symbols
    stable.
    
    Pass various defines like -D__DATE__=redacted to the wheel build to avoid
    getting timestamps into the binaries.
    
    If the wheel is built locally, strip timestamps from the whl zip file.
    
    Upgrade wheel package used by piptool to a later version (old one would
    produce non-deterministic wheels on python3).
    
    Support "sha256" attribute in the "download_or_build_wheel" rule, and use
    it in repository_ctx.download (when downloading) or pass to "piptool build",
    where we will check that the built wheel digest matches the expected one.
    (Note that "piptool resolve" does not yet emit the "sha256" attribute to
    the checked-in requirements.bzl, that will be done in a follow-up change).
    
    The build can still be non-deterministic when executed on different
    machines, since it's not hermetic.  But when built inside a container,
    it should be.

---

    Support generating stable sha256s from "piptool resolve"
    
    Add a new "digests" boolean attribute to pip_import.  If true, compute
    sha256 digests from all wheels after resolving.  This is done as follows:
    
      - If the wheel that gets bulit during "piptool resolve" has the same sha256
        that is already stored in the current version of the to-be-regenerated
        requirements.bzl, then use that as the sha256.  This is an optimization
        that speeds up sha256 resolution in the incremental resolve case where the
        wheel digest is the same whether it's built in isolation or as part of a
        non-isolated resolve operation.
    
      - Else, build the wheel in isolation (in a stable path to get deterministic
        debug symbols, using the correct additional_buildtime_env, using a clean
        PYTHONPATH that has just additional_buildtime_deps and their transitive
        runtime deps installed, etc.).  If the wheel build ends up just downloading
        a prebuilt wheel, use the sha256 of that without further stripping.
    
      - Else, if the wheel gets built locally, strip timestamps from the built
        zip file.  If the resulting stripped wheel has the same sha256 as in the
        current version of the to-be-generated requirements.bzl, then use that as
        the sha256.  This is an optimization for the incremental resolve cas where
        the wheel digest is different when built in isolation, but is not changing
        as part of the current resolve cycle.
    
      - Else, we have a wheel whose sha256 changed from the one produced during
        non-isolated resolve, and it doesn't match the current sha256 stored in
        the current version of requirements.bzl.  As a last wall of defence, build
        the wheel again, and make sure that we get the same sha256 from the second
        build.  If this check fails, then the wheel build is not deterministic.
        (I have not yet come across any wheels where this would happen.)

therc commented:
Sorry to invade a random pull request (I can't create an issue, either), but are there plans to send this repository's changes upstream? I saw Conrado's talk and chatted with him in person. I have been experimenting with several Python rule sets and assorted forks. This is the most comprehensive in handling bizarre packages. I'm not asking for a roadmap; only wondering if this is going to be a permanent fork. Thanks for the great work!

scele commented:
> Sorry to invade a random pull request

No problem at all! :)

> but are there plans to send this repository's changes upstream?

I'd like to upstream these changes, but it's a lot of work since this is a pretty significant rewrite.. So no, not planning this to be a permanent fork, it's just a matter of finding the time and energy to contribute back to upstream. :) :/

Glad you're finding it useful!

scele committed:
Refactor generated requirements.bzl layout and make overriding attributes easier

Previously, we would generate a bunch of macros and lookup tables to the
checked-in requirements.bzl that is generated by piptool.py during
@pip_deps//:update.  To clean things up (and to make it easier to make
changes to these macros), generate a simple "wheels" dict instead, which
contains all nexessary information about the resolved wheels.  Move the
macros to requirements.bzl.tpl instead, which is generated at runtime
and not checked into the user's repository.

Remove dirty package targets, since we don't need them anymore.  Rename
"extract_wheels" to "extract_wheel" and only accept one --whl parameter
in "piptool resolve", since the only use case for extracting multiple
wheels at once was the dirty wheel system.

Previously, we had to expose a new pip_import attribute for every
per-wheel customization that we whish to do: additional_runtime_deps,
additional_buildtime_env, patch_runtime, ...  This is a bit tedious to
maintain.  Instead, add a new "requirements_overrides" attribute, which
accepts a dictionary in the same format as the checked-in "wheels"
dictionary, and is merged together with that at analysis time.  This
allows users to override any attribute of any wheel, and as we add new
attributes to "download_or_build_wheel" and "extract_wheel" rules, they
are automatically overridable by the new mechanism.

Remove unnecessary templating from update.sh, and just generate the
update command directly from the repository rule.

scele committed:
Support extras and "target" argument in pip_version_proxy's requirement macro

Generate new select() targets for extras, and also allow user to pass additional
targets by adding "additional_targets" in the "requirements_overrides" dict.
This should remove the need for "requirement_repo" macro, since the pip_version_proxy
would now correctly proxy through things like:

   requirement("tensorflow", target = "cpp")

Since these new targets that we use in select() are not necessary py_libraries,
use native.alias instead of py_library targets.

"binary" attribute is still left unsupported by the version proxy.

scele committed:
Use standard patching attributes

Rename "patch_runtime" to "patches", and add other standard patching
attributes.

scele committed:
Move build_wheel logic from skylark into piptool

We will need to trigger the "build single wheel in isolation" logic
also from "piptool resolve" when generating stable sha256s later.

scele committed:
Small fix for macOS

scele committed:
Make wheel build deterministic

Use a fixed --build dir when invoking "pip wheel", to make debug symbols
stable.

Pass various defines like -D__DATE__=redacted to the wheel build to avoid
getting timestamps into the binaries.

If the wheel is built locally, strip timestamps from the whl zip file.

Upgrade wheel package used by piptool to a later version (old one would
produce non-deterministic wheels on python3).

Support "sha256" attribute in the "download_or_build_wheel" rule, and use
it in repository_ctx.download (when downloading) or pass to "piptool build",
where we will check that the built wheel digest matches the expected one.
(Note that "piptool resolve" does not yet emit the "sha256" attribute to
the checked-in requirements.bzl, that will be done in a follow-up change).

The build can still be non-deterministic when executed on different
machines, since it's not hermetic.  But when built inside a container,
it should be.

scele committed:
Support generating stable sha256s from "piptool resolve"

Add a new "digests" boolean attribute to pip_import.  If true, compute
sha256 digests from all wheels after resolving.  This is done as follows:

  - If the wheel that gets bulit during "piptool resolve" has the same sha256
    that is already stored in the current version of the to-be-regenerated
    requirements.bzl, then use that as the sha256.  This is an optimization
    that speeds up sha256 resolution in the incremental resolve case where the
    wheel digest is the same whether it's built in isolation or as part of a
    non-isolated resolve operation.

  - Else, build the wheel in isolation (in a stable path to get deterministic
    debug symbols, using the correct additional_buildtime_env, using a clean
    PYTHONPATH that has just additional_buildtime_deps and their transitive
    runtime deps installed, etc.).  If the wheel build ends up just downloading
    a prebuilt wheel, use the sha256 of that without further stripping.

  - Else, if the wheel gets built locally, strip timestamps from the built
    zip file.  If the resulting stripped wheel has the same sha256 as in the
    current version of the to-be-generated requirements.bzl, then use that as
    the sha256.  This is an optimization for the incremental resolve cas where
    the wheel digest is different when built in isolation, but is not changing
    as part of the current resolve cycle.

  - Else, we have a wheel whose sha256 changed from the one produced during
    non-isolated resolve, and it doesn't match the current sha256 stored in
    the current version of requirements.bzl.  As a last wall of defence, build
    the wheel again, and make sure that we get the same sha256 from the second
    build.  If this check fails, then the wheel build is not deterministic.
    (I have not yet come across any wheels where this would happen.)

scele committed:
Delete stale comment

scele committed:
Remove use of requirement attribute

scele committed:
Export wheel from the wheel repo

scele committed:
Remove dead code

scele committed:
Update examples

scele committed:
Update piptool.par

mirandaconrado commented on rules_python/piptool.py:
```diff
@@ -551,7 +566,10 @@ def digest(fname):
-  ordering = build_dep_graph(args)
+  print("Generating %s from %s..." % (args.output, " and ".join(args.input)))
+
+  build_info = ast.literal_eval(args.build_info or '{}')
```
`literal_eval` is generally a bad idea. Specially as I don't recall what it's supposed to be here.

scele replied:
Added a comment and changed to json:
```
  # Parse build_info - this is the contents of "requirements_overrides" attribute
  # passed by the user, serialized to json.
  build_info = json.loads(args.build_info or '{}')
```

scele replied:
BTW, `literal_eval` is not that bad (compared to `eval`), and I use `literal_eval` anyway for parsing the contents of `requirements.bzl` since that's a starlark dict, not json (trailing commas and comments are not supported by json parser).

mirandaconrado commented on rules_python/piptool.py:
```diff
@@ -646,6 +664,54 @@ def transitive_deps(wheel, extra=None, collected=None):
+          continue
+
+      build_deps = set().union(*[
+        {whl_map[d].path()} | {whl_map[w].path() for w in transitive_deps(whl_map[d])}
```
Ugh... What's going on here? Can we expand from a single line for easier understanding?

scele replied:
Cleaned up.

mirandaconrado commented on rules_python/piptool.py:
```diff
@@ -717,8 +785,8 @@ def whl_library(wheel):
 parser.add_argument('--directory', action='store', default='.',
                     help=('The directory into which to put .whl files.'))
 
-parser.add_argument('--python', action='store',
```
Why removing?

scele replied:
We used to write the python interpreter into the generated checked-in requirements.bzl. But now that I simplified its format to being just `wheels = { <json> }`, we don't need this arg anymore.

mirandaconrado commented on rules_python/piptool.py:
```diff
@@ -143,21 +168,56 @@ def build_wheel(distribution,
+  if sha256 and computed_sha256 != sha256:
+    # If the user supplied an expected sha256, we must match it.
+    print("Built wheel %s digest %s does not match expected digest %s." % (wheel.path(), computed_sha256, sha256))
+    sys.exit(1)
```
Should we error on this or give a warning? I can see people getting blocked because of this.

scele replied:
Good idea, done.

mirandaconrado commented on python/pip.bzl:
```diff
@@ -120,6 +98,14 @@ pip_import = repository_rule(
+    if "requirements_overrides" in kwargs:
+        # Overrides are serialized to string and passed to the rule, since
+        # rules cannot have deep dicts.
+        kwargs["requirements_overrides"] = str(kwargs["requirements_overrides"])
```
Should we use a more standard format like json to ensure we don't have arbitrary types going around?

scele replied:
Done.

mirandaconrado commented on python/pip.bzl:
```diff
@@ -13,84 +13,67 @@
+    cmd += [
+        "--output=%s" % str(ctx.path(ctx.attr.requirements_bzl)),
+        "--output-format=download",
+        "--directory=%s" % str(ctx.path("build-directory")),
```
Is there any link to docs on this? Not sure how it works. I didn't know we could get the final file path in the repo from a rule.

scele replied:
From https://docs.bazel.build/versions/master/skylark/lib/repository_ctx.html#path:
> If the path is relative, it will resolve relative to the repository directory.

mirandaconrado commented:
LGTM but honestly I had quite a hard time following everything that was done. Maybe bc I haven't looked it in a while.

scele committed:
Update examples

scele committed:
Latest keras-preprocessing seems to be missing pandas dep

scele committed:
Pass "requirements_overrides" to piptool as json instead of serialized dict

scele committed:
Call ctx.report_progress to indicate extract/build steps

scele committed:
Render "build_deps" instead of "transitive_runtime_deps" into generated requirements.bzl

Figuring out buildtime deps is a bit tricky, since we need to take the
"additional_buildtime_deps" override from the user, and collect all
transitive runtime dependencies from those distributions, while also
taking into account any "additional_runtime_deps" provided by the user.
To simplify the logic, calculate the correct set of build deps already
at resolve time, and write as "build_deps" arttibute to the generated
requirements.bzl.

scele committed:
Print pip command on failure

scele committed:
Print a warning, instead of failing, on built wheel digest mismatch

scele committed:
Add a few comments

scele committed:
Remove dead code

scele committed:
Update piptool
